# Pre-Lecture HW

### Question 1: 

In [3]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
villagers = pd.read_csv(url)
villagers.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

In [4]:
url2 = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
titanic = pd.read_csv(url2)
titanic.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

### Question 2: 

In [5]:
print(titanic.shape)
titanic

(891, 15)


survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [6]:
print(villagers.shape)
villagers

(391, 11)


row_n        id      name  gender    species birthday personality  \
0        2   admiral   Admiral    male       bird     1-27      cranky   
1        3   agent-s   Agent S  female   squirrel      7-2       peppy   
2        4     agnes     Agnes  female        pig     4-21        uchi   
3        6        al        Al    male    gorilla    10-18        lazy   
4        7   alfonso   Alfonso    male  alligator      6-9        lazy   
..     ...       ...       ...     ...        ...      ...         ...   
386    475    winnie    Winnie  female      horse     1-31       peppy   
387    477  wolfgang  Wolfgang    male       wolf    11-25      cranky   
388    480      yuka      Yuka  female      koala     7-20      snooty   
389    481      zell      Zell    male       deer      6-7        smug   
390    483    zucker    Zucker    male    octopus      3-8        lazy   

                song    phrase            full_id  \
0         Steep Hill   aye aye   villager-admiral   
1            DJ K.K.  sidekick   villager-agent-s   
2         K.K. House   snuffle     villager-agnes   
3         Steep Hill   Ayyeeee        villager-al   
4        Forest Life  it'sa me   villager-alfonso   
..               ...       ...                ...   
386         My Place    hay-OK    villager-winnie   
387        K.K. Song   snarrrl  villager-wolfgang   
388     Soulful K.K.   tsk tsk      villager-yuka   
389         K.K. D&B     pronk      villager-zell   
390  Spring Blossoms     bloop    villager-zucker   

                                                   url  
0    https://villagerdb.com/images/villagers/thumb/...  
1    https://villagerdb.com/images/villagers/thumb/...  
2    https://villagerdb.com/images/villagers/thumb/...  
3    https://villagerdb.com/images/villagers/thumb/...  
4    https://villagerdb.com/images/villagers/thumb/...  
..                                                 ...  
386  https://villagerdb.com/images/villagers/thumb/...  
387  https://villagerdb.com/images/villagers/thumb/...  
388  https://villagerdb.com/images/villagers/thumb/...  
389  https://villagerdb.com/images/villagers/thumb/...  
390  https://villagerdb.com/images/villagers/thumb/...  

[391 rows x 11 columns]

Observation represents rows within a dataset. Generally, observation focuses on one entity of a dataset and all the information about that entity. In the animal crossing dataset though, one observation would be the entity "Agent S", and that represents all the information about "Agent S". Its birthday, species, phrase, etc... 

Variable represents columns of the dataset, where each column would store the same type of information, but information regarding different entities. For example, "id" is a variable that stores identification of every villager. It's the same type of data, but data from different observations. 

### Question 3 

In [7]:
titanic.describe(include='all')

survived      pclass   sex         age       sibsp       parch  \
count   891.000000  891.000000   891  714.000000  891.000000  891.000000   
unique         NaN         NaN     2         NaN         NaN         NaN   
top            NaN         NaN  male         NaN         NaN         NaN   
freq           NaN         NaN   577         NaN         NaN         NaN   
mean      0.383838    2.308642   NaN   29.699118    0.523008    0.381594   
std       0.486592    0.836071   NaN   14.526497    1.102743    0.806057   
min       0.000000    1.000000   NaN    0.420000    0.000000    0.000000   
25%       0.000000    2.000000   NaN   20.125000    0.000000    0.000000   
50%       0.000000    3.000000   NaN   28.000000    0.000000    0.000000   
75%       1.000000    3.000000   NaN   38.000000    1.000000    0.000000   
max       1.000000    3.000000   NaN   80.000000    8.000000    6.000000   

              fare embarked  class  who adult_male deck  embark_town alive  \
count   891.000000      889    891  891        891  203          889   891   
unique         NaN        3      3    3          2    7            3     2   
top            NaN        S  Third  man       True    C  Southampton    no   
freq           NaN      644    491  537        537   59          644   549   
mean     32.204208      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
std      49.693429      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
min       0.000000      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
25%       7.910400      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
50%      14.454200      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
75%      31.000000      NaN    NaN  NaN        NaN  NaN          NaN   NaN   
max     512.329200      NaN    NaN  NaN        NaN  NaN          NaN   NaN   

       alone  
count    891  
unique     2  
top     True  
freq     537  
mean     NaN  
std      NaN  
min      NaN  
25%      NaN  
50%      NaN  
75%      NaN  
max      NaN

In [8]:
titanic['deck'].value_counts()

deck
C    59
B    47
D    33
E    32
A    15
F    13
G     4
Name: count, dtype: int64

### Question 4

The total count for the 'age' variable is 714 and there are 891 rows for the data frame. The discrepency is probably due to data being missing or the data was never recorded. 
For the 'deck' varible, 688 pieces of data are missing, and this is probably due to the reason that only the survivors were able to give information on which deck they were on at the time of the collision, and many of the survivors probably don't remember either. Only the observations with 'survived' as '1', has data in their 'deck' variable (working with the few rows that are shown in jupyter). 

### Question 5

Attribute is something inherent to the object, it's an characteristic. It stores data for the object and parentheses aren't needed because it isn't a function call, it is just retrieving a value. 

Method is a type of function, more specifically it's a function that only works with that particular object. Method evaluates / performs actions on arguments given to it (not all methods or functions require arguments though). 

# Pre-Lecture HW Chat Summary 

https://chatgpt.com/share/2289b0c2-8178-47b0-9a52-bd86b069da3b

We began by discussing a dataset of villagers, specifically from the URL you provided. After an initial issue accessing the dataset, we explored key concepts such as observations (rows representing individual villagers) and variables (columns representing attributes like species or personality).

You asked for code to summarize the dataset, and I explained how .describe() provides summaries for both numerical and categorical columns. When you encountered an issue with df.describe(include=number), I clarified that number should be replaced with either np.number (if using NumPy) or ['float64', 'int64'] for numeric columns using only Pandas.

We also compared .describe() with .value_counts(), noting that .describe() provides general statistics, while .value_counts() focuses on frequency counts of unique values in a single column.

Finally, we covered the difference between attributes (characteristics or properties of an object) and methods (functions that perform actions on an object).

# Post-Lecture HW

### Question 6

Count: Number of entries, not including 'NaN'
Mean: Average value of the entries 
Standard deviation (std): It measures how "spread out" or how much the data deviate from the mean
Minimum (min): Smallest value in a variable 
25th Percentile (25%): The value where 25% of the data falls, also called the first quartile
50th Percentile (50%): The value where half of the data falls, also known as the median
75th Percentile (75%): The value where 75% of the data fals, also called the third quartile
Maximum (max): Largest value in a variable

### Question 7

#### 1.

When someone would want to remove only the NaN value and nothing else. del df['col'] would delete the entire column. Also df.dropna() gives you more customizability, as you can also use it on rows, while del df['col'] can't be. 

#### 2.

When someone would just want to get rid of an entire column, and not leave data behind. 

#### 3. 

When someone would like to clense their data frame, it will be useful to have del df['col'] getting rid of all the irrelevant data, and then using df.dropna() to get rid of the rows/columns with NaN. 

#### 4.

In [9]:
titanic # before

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [ ]:
del titanic['deck'] #isn't retre

In [15]:
titanic.dropna(inplace=True)
titanic

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
885         0       3  female  39.0      0      5  29.1250        Q   Third   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male  embark_town alive  alone  
0      man        True  Southampton    no  False  
1    woman       False    Cherbourg   yes  False  
2    woman       False  Southampton   yes   True  
3    woman       False  Southampton   yes  False  
4      man        True  Southampton    no   True  
..     ...         ...          ...   ...    ...  
885  woman       False   Queenstown    no  False  
886    man        True  Southampton    no   True  
887  woman       False  Southampton   yes   True  
889    man        True    Cherbourg   yes   True  
890    man        True   Queenstown    no   True  

[712 rows x 14 columns]

I decided to delete the 'deck' column right away because there are so many missing entries, so if I use titanic.dropna() the entire dataset might as well be gone. So after using del titanic['deck'] I used titanic.dropna(inplace=True) to modify the dataset directly instead of having a copy which I could have done. 

## Question 8

#### 1. 

In [10]:
titanic.groupby('alone')['survived'].describe()

count      mean       std  min  25%  50%  75%  max
alone                                                    
False  354.0  0.505650  0.500676  0.0  0.0  1.0  1.0  1.0
True   537.0  0.303538  0.460214  0.0  0.0  0.0  1.0  1.0

.groupby('col1')['col2'] functions by grouping all the unique entries of 'col1' together. And then it will only look at 'col2' and the data in 'col2' would be sorted based on the unique entries of 'col1'. Finally with .describe() it will generate all the statistical analysis values for the user to use. 

So in my example, I wanted to look at whether being alone on the titanic would impact your chances at survival. So I used .groupby('alone') to separate the loners from the groups of people. Then have .describe() analyze the ['survived'] in the loners and the groups. And now I can see the loners had a significantly worse chance of survival than the groups. 

#### 2.

Using .describe() initially is just listing out data for you. It doesn't mean anything, instead it's giving you facts that someone could use to derive a meaning. But with .groupby('col1')['col2'].describe() two different parts are being compared, that process of comparison leads to patterns and more information gained. Using my example above, I was able to find out that groups of people had a significantly higher chance of survival than people that were alone. Which, if we didn't use .groupby('col1')['col2'] would be impossible to see with just the fact that 38% of people survived from the collison. 

#### 3.

For some problems Chat and Google performed to the same level of proficiency, but for Google I have to actively search out answers, click on different links, and sometimes the link I fist click on doesn't resolve my issue so I have to do it again. For Chat it comes down to convenience, when I ask Chat something they answer me right away. And for these coding problems Chat is pretty good at detecting what's wrong. Maybe there is a bias due to the fact that I have been using the same conversation with Chat for my HW so Chat is in the STA130 frame of mind. Still it is way easier to copy and paste the problem into Chat than to go on Google to try to search for an answer. 

### Question 9

Somewhat. I didn't have any experience with pandas or jupyter or github so I was super confused. Especially after the first tutorial where nothing was really explained to me and was just thrown into the deepend. But Chat really helped me, any questions Chat would answer to a good degree of satisfaction. 

# Post-Lecture HW Chat Summary


https://chatgpt.com/share/c6b90107-15f9-46c6-9d1e-b63769449d9c

Dataset: We worked with the Titanic dataset from the URL: https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv. However, when loading the dataset, there were a few issues with URLs and accessing data, which were resolved by using another URL: https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv.

Exploration of .describe() Method:

We discussed the .describe() method in Pandas, which provides summary statistics for each column, including count, mean, std, min, max, and percentiles (25%, 50%, 75%).
Standard Deviation: We clarified the meaning of standard deviation as a measure of how spread out the values are from the mean.

Deleting Columns (del df['col']):

You learned that the del statement deletes entire columns from a DataFrame.
We discussed why someone might use del before applying df.dropna(), for example, to remove irrelevant columns before handling missing values.
Handling Missing Data (df.dropna()):

You learned how df.dropna() works to remove rows or columns containing missing values (NaNs).
You encountered issues, such as KeyError, due to attempting to delete a column that didn’t exist.
Errors and Debugging:

Several errors, including NameError, HTTPError, and KeyError, were encountered and resolved.
The NameError occurred due to missing imports, and the solution was to import the necessary libraries (import pandas as pd).
The KeyError happened when trying to access columns that did not exist in the dataset, which was resolved by checking the DataFrame’s column names using print(df.columns).
Groupby and Describing Statistics:

We explored how to use the groupby() function with describe() to group data by certain columns (like class) and get descriptive statistics for other columns (like age).
We fixed issues related to the incorrect use of group_by (should be groupby) and demonstrated correct usage with the Titanic dataset.
Error: Column Does Not Exist:

When attempting to access a non-existent column (e.g., 'ALONE'), you encountered a KeyError. We resolved this by ensuring the column exists using if 'column' in df.columns.